<head>
  <title>Blixem - Release History</title>
</head>


<body>
<h1>Blixem - Release History</h1>

  <h3>v4.4</h3>
  <ul>
    <li>There is a new --zoom-range argument, which allows the caller to specify the initial big picture range.</li>
    <li>Reference sequence coordinates can now be specified in the FASTA definition line in the sequence file; the format must be ">name start end".</li>
    <li>Fixed a bug where introns that are partially out of range were not being shown.</li>
    <li>Fixed a bug where short reads were incorrectly being squashed onto the same line in 'squash matches' mode when their sources are different.</li>
    <li>Fixed a bug where the initial exon view height was sometimes being calculated incorrectly.</li>
    <li>Fixed a bug where starting Dotter fails if the fetch command returns a string such as "no match".</li>
  </ul>


  <h3>v4.2</h3>
  <ul>
    <li>Blixem now supports fetching of short-read data from BAM files.</li>
    <li>Blixem command-line arguments have changed.</li>
  </ul>
  
  
  <h3>v4.1.14</h3>
  <ul>
    <li><b>Short reads</b>: The 'short_read' feature type is now supported.</li>
    <li><b>Coverage</b>: A new 'coverage view' has been added to show coverage depth - it can be enabled via the View dialog.</li>
    <li><b>Sorting</b>: The sort-by drop-down box has been replaced with a Sort dialog, allowing sorting by multiple columns - open it with the Sort button on the toolbar.  The 'invert sort order' option has been moved from the Settings dialog to the Sort dialog.</li>
    <li><b>Search by column</b>: When grouping or finding sequences, you can now search on data in different columns, not just the Name column.  See the Group and Find dialogs.</li>
    <li><b>Groups</b>: You can now hide all sequences that are not in a group.  See the Groups dialog.</li>
  </ul>


  <h3>v4.1.13</h3>
  <ul>
    <li>Replaced the Help dialog with HTML help pages.</li>
  </ul>


  <h3>v4.1.12</h3>
  <ul>
    <li>You can now change the active reading frame by clicking in a list header (as well as in the list itself).</li>
  </ul>


  <h3>v4.1.11</h3>
  <ul>
    <li>The 'Source' column is now enabled by default.  It is also sortable and its contents are now shown in the feedback area.</li>
    <li>Column widths can now be specified in the Blixem config file (which is supplied via the <code>-c</code> command line argument).</li>
  </ul>


  <h3>v4.1.10</h3>
  <ul>
    <li>Fixed a bug with populating optional data for protein variants.</li>
  </ul>


  <h3>v4.1.9</h3>
  <ul>
    <li>Printing now scales to fit the page.</li>
    <li>Fixed a bug where splice sites were incorrectly being shown at termini.</li>
  </ul>


  <h3>v4.1.5</h3>
  <ul>
    <li><b>Close all Dotters</b>: There is a new menu option to close all Dotters started from the current Blixem. Right-click in Blixem to see the menu.</li>
  </ul>

  
  <h3>v4.1.4</h3>
  <ul>
    <li><b>Variations</b>: SNPs, insertions, deletions etc. can be highlighted in the reference sequence, and shown in a 'variations track' above the reference sequence. See the Settings dialog or double-click the reference sequence to show/hide the variations track. Double-click a variation to open its URL. This is an experimental feature and the display is quite crude - please give us feedback on how you would like variations to be displayed, particularly where they overlap.</li>
    <li><b>PolyA tails</b>: PolyA tails can be shown and polyA signals highlighted in the reference sequence. See the Settings dialog.</li>
    <li><b>CDS/UTR regions</b>: Exons are now separated into CDS and UTR regions: CDS regions are coloured green and UTR red. Don't forget, you can quickly bump/un-bump the exon view by pressing 'b'.</li>
    <li><b>Moused-over item feedback</b>: There is now an additional feedback area on the toolbar that shows information about the currently moused-over item.</li>
    <li><b>Status bar</b>: There is now a status bar at the bottom of the window, which shows non-critical messages and warnings.</li>
    <li><b>Grid scale</b>: The grid scale now has a finer level of granularity; you can set %ID per cell to as low as 0.1 (see the Settings dialog).</li>
    <li><b>New Dotter</b>: Dotter has had a re-vamp - check the Dotter Help page for more info.</li>
  </ul>

</body>
